### PR TITLE
[beetmover] Add GCS support

### PR DESCRIPTION
**Pull request by hneiva**

### CODE REVIEW

**Comment by hneiva** [comment]
Haven't added any tests yet, want to get extra pairs of eyes on it before spending more time on it.

File: beetmoverscript/docker.d/worker.yml [hunk 1]
```diff
@@ -30,6 +30,8 @@ bucket_config:
     $match:
       'COT_PRODUCT == "firefox" && (ENV == "dev" || ENV == "fake-prod")':
         dep:
+          gcs_credentials: { "$eval": "GCS_CREDENTIALS" }
+          fail_task_on_gcs_error: False
```
Comment (escapewindow) [inline beetmoverscript/docker.d/worker.yml]: I'm not sure I love this layout.

a) it assumes that S3 is the main way of doing things, and GCS is a secondary way of doing things, and those assumptions are baked into the code.
b) when we're ready to cut over so that GCS is the primary way of doing things, we may need to land a very large code change. (I'm guessing -- are the buckets named the same in GCS as S3? If so, it's a smaller delta, but still.)

I'd probably look something like

```
clouds:
    s3:
       credentials: ...
       enabled: True
       fail_on_error: True
       ... # if url_prefix verification etc are cloud-specific, list them here rather than in the shared config space
    gcs:
       ...
```

This takes more up-front work, but if we want to require both clouds upload successfully, that's a single boolean flip. If we are ready to cut over to GCS and stop uploading to S3, that's another boolean flip.

This may be a me thing: I prefer to make things more configuration-driven than hardcoded code-logic driven.

File: beetmoverscript/src/beetmoverscript/script.py [hunk 2]
```diff
@@ -303,23 +308,69 @@ async def async_main(context):
 
     validate_task_schema(context)
 
+    # determine the task bucket and action
+    #   Note: bucket here is the config key under bucket_config, for aws/gcs
+    #         platform "bucket" name use get_bucket_name()
+    context.bucket = get_task_bucket(context.task, context.config)
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/script.py:321]: While you're in here refactoring, you may want to keep in #171, which would get us to stop using the `context` object from scriptworker, and start passing around dictionaries directly. Definitely scope creep, but it's definitely more of a bug than a feature that we import scriptworker in our scriptworker-scripts.

File: beetmoverscript/src/beetmoverscript/script.py [hunk 3]
```diff
@@ -303,23 +308,69 @@ async def async_main(context):
...
+    fp.close()
+    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = fp.name
+
+
+# get_gcs_client {{{1
+def set_gcs_client(context):
+    client = Client()
+    product = get_product_name(context.release_props["appName"].lower(), context.release_props["stage_platform"])
+    bucket = client.bucket(get_bucket_name(context, product))
+    try:
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/script.py]: We probably want to make this configurable as well.
Am I mistaken, or can we set gcs credentials, mark the fail-on-gcs-error as `true`, and yet we'll completely skip gcs uploads if there's a bucket config issue? That sounds like a bug.

Comment (hneiva) [reply beetmoverscript/src/beetmoverscript/script.py]: Down side of not having tests with the code. Nice catch!

File: beetmoverscript/docker.d/worker.yml (hunk 1 above)
Comment (hneiva) [reply beetmoverscript/docker.d/worker.yml]: Heh that's how I started with my changes. But after realizing how tightly coupled with `bucket_config` it is, I decided to move the GCS config in with s3 stuff.

It's fairly easy to move it out, but it will still require more work to be able to cutoff S3.
Should I add the ability for us to cutoff already, or should we just support adding GCS for now and cutoff later?

File: beetmoverscript/src/beetmoverscript/script.py (hunk 2 above)
Comment (hneiva) [reply beetmoverscript/src/beetmoverscript/script.py:321]: I think we should fix this in a different PR. It touches pretty much every function in beetmoverscript.

File: beetmoverscript/docker.d/worker.yml (hunk 1 above)
Comment (escapewindow) [reply beetmoverscript/docker.d/worker.yml]: This probably depends on how urgent the request is. If we have time, then doing the work ahead of time makes sense to me. If getting GCS populated sooner is urgent, we might have to land this in phases.

Comment (ahal) [reply beetmoverscript/docker.d/worker.yml]: I'd guess we have a month or two before this becomes super urgent.

File: beetmoverscript/src/beetmoverscript/script.py [hunk 4]
```diff
@@ -303,23 +308,71 @@ async def async_main(context):
...
+
+    gcs_creds = context.config["bucket_config"][context.bucket].get("gcs_credentials")
+    if type(gcs_creds) is str and len(gcs_creds) > 0:
+        setup_gcs_credentials(gcs_creds)
+        # TODO: should we load release_props in async_main instead of each action?
+        #   Needed for bucket lookup
+        context.release_props = get_release_props(context)
+        set_gcs_client(context)
+    else:
+        log.info("No GCS credentials found, skipping")
```
Comment (ahal) [inline beetmoverscript/src/beetmoverscript/script.py]: Can we tell if the request is going to be using GCS and make this an error if it's missing?

**Review by ahal — CHANGES_REQUESTED** [review]
This looks reasonable to me, though I don't have any familiarity with beetmoverscript. Re: whether to implement the S3 cutoff ability now or later, I'd vote now if you don't think it would take unreasonably long.

I think it'll give us more confidence when it's time to make the switch.

**Comment by hneiva** [comment]
Latest push should implement enabling/disabling a cloud.

File: beetmoverscript/requirements/test.in [hunk 5]
```diff
@@ -3,7 +3,7 @@
 black
 check-manifest
 coverage
-flake8
+flake8==5.0.4
```
Comment (escapewindow) [inline beetmoverscript/requirements/test.in]: Could we add a comment for why this is pinned?
Ideally we link to issues, and we can remove the pin when those issues are resolved.

File: beetmoverscript/src/beetmoverscript/data/config_schema.json [hunk 6]
```diff
@@ -6,7 +6,7 @@
         "taskcluster_scope_prefixes",
         "aiohttp_max_connections",
         "zip_max_file_size_in_mb",
-        "bucket_config"
+        "clouds"
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/data/config_schema.json]: Should we add url_prefix here too?

[EDIT] on second glance, it looks like url_prefix at the top level may be left over.

File: beetmoverscript/docker.d/worker.yml [hunk 7]
```diff
@@ -25,250 +25,465 @@ taskcluster_scope_prefixes:
...
+      release: 'https://ftp.stage.mozaws.net'
+    'COT_PRODUCT == "mobile" && ENV == "prod"':
+      nightly: 'https://archive.mozilla.org'
+      release: 'https://archive.mozilla.org'
+    'COT_PRODUCT == "mozillavpn" && (ENV == "dev" || ENV == "fake-prod")':
+      dep: 'https://ftp.stage.mozaws.net'
+    'COT_PRODUCT == "mozillavpn" && ENV == "prod"':
+      release: 'https://archive.mozilla.org'
+
+clouds:
```
Comment (escapewindow) [inline beetmoverscript/docker.d/worker.yml:61]: Nice.

I have not done a precise, in-depth review of these changes -- if there are any questions, it may be good to write a quick script that looks at the old json-e configs and new json-e configs, and verifies that the aws configs have not changed (other than layout), and possibly that the gcs configs appear to match the aws configs.

I know that sounds like a pain, but IMO it's way preferable over discovering the bustage when a critical release is broken.

File: beetmoverscript/requirements/local.in [hunk 8]
```diff
@@ -1,3 +1,7 @@
 -r test.in
 
 tox
+
+# virtualenv created some conflicts with importlib-metadata after 20.16.3
+# https://github.com/pypa/virtualenv/pull/2394
```
Comment (escapewindow) [inline beetmoverscript/requirements/local.in]: 👍 
Is there a reason why we were using py36?

File: beetmoverscript/docker.d/worker.yml [hunk 9]
```diff
@@ -25,250 +25,465 @@ taskcluster_scope_prefixes:
...
-      'COT_PRODUCT == "mozillavpn" && ENV == "prod"':
-        release:
-          credentials:
-            id: { "$eval": "RELEASE_ID" }
-            key: { "$eval": "RELEASE_KEY" }
-          buckets:
-            vpn: 'net-mozaws-prod-delivery-archive'
-          url_prefix: 'https://archive.mozilla.org'
-        dummy: {}
+url_prefix:
```
Comment (escapewindow) [inline beetmoverscript/docker.d/worker.yml:29]: Hm. We specify url_prefix here, and also in clouds.aws, and the schema only lists the latter.
Is this url_prefix in use?

File: beetmoverscript/src/beetmoverscript/data/config_schema.json [hunk 10]
```diff
@@ -30,34 +30,28 @@
         "zip_max_file_size_in_mb": {
             "type": "number"
         },
-        "bucket_config": {
+        "clouds": {
             "type": "object",
             "additionalProperties": {
                 "type": "object",
-                "additionalProperties": false,
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/data/config_schema.json:37]: Is this right? You have `additionalProperties` as a dict that contains the `properties`, rather than as a boolean.

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    if hasattr(context, "gcs_client") and context.gcs_client:
+        os.remove(os.environ["GOOGLE_APPLICATION_CREDENTIALS"])
+
+
+def setup_gcloud(context):
+    gcs_creds = get_credentials(context, "gcloud")
+    if type(gcs_creds) is str and len(gcs_creds) > 0:
+        setup_gcs_credentials(gcs_creds)
+        # TODO: maybe we should load release_props in async_main instead of each action?
+        #   Needed for bucket lookup
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/gcloud.py]: 👍

File: beetmoverscript/src/beetmoverscript/script.py [hunk 12]
```diff
@@ -517,7 +529,7 @@ def generate_system_addons_balrog_manifest(context):
             addon_data = get_addon_data(filepath)
             addon_name = addon_data["name"]
             addon_version = addon_data["version"]
-            addon_url = "{prefix}/{s3_key}".format(prefix=get_bucket_url_prefix(context), s3_key=destinations[0])
+            addon_url = "{prefix}/{s3_key}".format(prefix=get_url_prefix(context), s3_key=destinations[0])
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/script.py:550]: I kind of want us to rename `s3_key` if it's not s3-specific.

File: beetmoverscript/src/beetmoverscript/script.py [hunk 13]
```diff
@@ -587,10 +599,31 @@ def enrich_balrog_manifest(context, locale):
 # retry_upload {{{1
 async def retry_upload(context, destinations, path):
     """Manage upload of `path` to `destinations`."""
-    uploads = []
+    cloud_uploads = {
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/script.py]: We can build this from config, but that could be overengineering :)

File: beetmoverscript/src/beetmoverscript/task.py [hunk 14]
```diff
@@ -78,7 +78,7 @@ def get_task_bucket(task, script_config):
     if re.search("^[0-9A-Za-z_-]+$", bucket) is None:
         messages.append("Bucket {} is malformed".format(bucket))
 
-    if bucket not in script_config["bucket_config"]:
+    if bucket not in {*script_config["clouds"]["aws"].keys(), *script_config["clouds"]["gcloud"].keys()}:
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/task.py]: I don't love this hardcode either, but probably not the worst.

**Review by escapewindow — CHANGES_REQUESTED** [review]
Looking good! I largely looked at the multi-cloud design, rather than logic specifics.
There are a few rough edges that I noticed. I have not in-depth looked at the logic; this is worth testing in a staging firefox try release at the least.

File: beetmoverscript/requirements/local.in (hunk 8 above)
Comment (hneiva) [reply beetmoverscript/requirements/local.in]: That was an oopsie on my side. I was talking to ahal about it and realized 3.6 isn't used for this anymore. 
So the base files (base.in, local.in, test.in) are py3.9.7 already, and then we support 3.8 on the .py38.txt files.
I'll clean it up

File: beetmoverscript/src/beetmoverscript/script.py (hunk 12 above)
Comment (hneiva) [reply beetmoverscript/src/beetmoverscript/script.py:550]: Yeah, in this case will likely be a dynamic function. I just haven't really looked into the balrog functions yet. But it's coming :)

File: beetmoverscript/src/beetmoverscript/data/config_schema.json (hunk 10 above)
Comment (hneiva) [reply beetmoverscript/src/beetmoverscript/data/config_schema.json:37]: It's some odd behavior of [json-schema additionalProperties](http://json-schema.org/understanding-json-schema/reference/object.html#additional-properties):

> The additionalProperties keyword is used to control the handling of extra stuff, that is, properties whose names are not listed in the properties keyword or match any of the regular expressions in the patternProperties keyword. By default any additional properties are allowed.
> 
> The value of the additionalProperties keyword is a schema that will be used to validate any properties in the instance that are not matched by properties or patternProperties. Setting the additionalProperties schema to false means no additional properties will be allowed.

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+            raise e
+        log.error(e)
+
+    # Only used for exception handling/logging
+    bucket = "<unset>"
+    try:
+        client = Client()
+        bucket = client.bucket(get_bucket_name(context, product, "gcloud"))
+        if not bucket.exists():
+            log.warning(f"GCS bucket {bucket} doesn't exit. Skipping GCS uploads.")
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/gcloud.py]: nit: `exist`

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+    if not candidates_blobs:
+        raise ScriptWorkerTaskException("No artifacts to copy from {} so there is no reason to continue.".format(candidates_prefix))
+
+    if releases_blobs:
+        log.warning("Destination {} already exists with {} keys".format(releases_prefix, len(releases_blobs)))
+
+    blobs_to_copy = {}
+
+    # Weed out RELEASE_EXCLUDE matches, but allow partners specified in the payload
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/gcloud.py:125]: If we plan on having S3 and GCS side-by-side for a lengthy amount of time, we could refactor common logic into helper functions to remove duplication. If we're planning on migrating fully, then this isn't necessary.

Not a blocker; noting for future improvements.

File: beetmoverscript/src/beetmoverscript/script.py [hunk 17]
```diff
@@ -167,8 +170,18 @@ async def push_to_partner(context):
...
+    if context.config["clouds"]["aws"][context.bucket]["enabled"]:
+        await push_to_releases_s3(context)
+
+    # GCS upload
+    if context.config["clouds"]["gcloud"][context.bucket]["enabled"]:
+        await push_to_releases_gcs(context)
+
+
+# push_to_releases_s3 {{{1
+async def push_to_releases_s3(context):
```
Comment (escapewindow) [inline beetmoverscript/src/beetmoverscript/script.py:184]: We're making enough changes in here that if we haven't already, it would be good to verify a staging fx release looks the same as a previous fx release. This is definitely something that makes the releng changelog; please be available for any beetmover issues during the next nightly / beta / release / esr.

**Review by escapewindow — APPROVED** [review]

File: beetmoverscript/src/beetmoverscript/script.py (hunk 17 above)
Comment (hneiva) [reply beetmoverscript/src/beetmoverscript/script.py:184]: Agreed. If we decide AWS should be supported long-term, we should refactor the s3 functionality into it's own module (like I did with gcloud).
Contemplating only releasing to prod after all-hands, as I'm off next week.

**Review by ahal — APPROVED** [review]
Lgtm!
